app/controllers/oaisys/pmh_controller.rb

```diff
@@ -2,25 +2,49 @@
 
 module Oaisys
   class PMHController < ApplicationController
+    layout 'oai_pmh_response.xml.builder'
+
     def bad_verb
     end
 
     def identify
+      expect_args(__method__, required: %i(), optional: %i(), exclusive: %i())
     end
 
     def list_sets
+      expect_args(__method__, required: %i(), optional: %i(), exclusive: %i(resumptionToken))
     end
 
     def list_metadata_formats
+      expect_args(__method__, required: %i(), optional: %i(identifier), exclusive: %i())
     end
 
     def list_records
+      expect_args(__method__, required: %i(metadataPrefix), optional: %i(from until set), exclusive: %i(resumptionToken))
     end
 
     def get_record
+      expect_args(__method__, required: %i(identifier metadataPrefix), optional: %i(), exclusive: %i())
     end
 
     def list_identifiers
+      expect_args(__method__, required: %i(metadataPrefix), optional: %i(from until set), exclusive: %i(resumptionToken))
+    end
+
+    private
+
+    def expect_args (verb, required:, optional:, exclusive:)
+      arguments = params.except('verb', 'controller', 'action').keys.map(&:to_sym)
+      expected_verb_arguments = required + optional + exclusive
+      unexpected_arguments = (arguments - expected_verb_arguments).present?
+      missing_required_arguments = (required - arguments).present?
+
+      if unexpected_arguments || missing_required_arguments
+        verb = verb.to_s.camelize
+        error_code = :badArgument
+        error_message = t('error_messages.illegal_or_missing_arguments')
+        render template: 'responses/error_response.xml.builder', locals: { verb: verb, error_code: error_code, error_message: error_message }
```

Comment:
If you move the views into `pmh/` instead of `responses/` (conventional, since that's the name of the controller), you should just be able to do `render :error_response`.

Broader issue (which I didn't think about before): this all works right now, but the second that we start implementing the method bodies (as in, fleshing out `list_metadata_formats`), what's going to happen if there are bad arguments is that `expect_args` is going to render this template and then return and _keep going_ in `list_metadata_formats`, which will either eventually crash from missing arguments or call `render` again and cause a double-render error.

So what we actually want to do is just _detect_ the error here and then, if we detect the error, raise an exception, and leave the actual setting of the error_message and error_code and rendering of an error response to a separate method that is only invoked if an exception is raised during execution.

I'd suggest creating a BadArgumentException type (and maybe a PMHException parent class that it inherits?) that wraps up the code and message. Raise that here, and then use https://apidock.com/rails/ActiveSupport/Rescuable/ClassMethods/rescue_from to trap the parent type and have it call the `render`. That way we reliably either continue executing from `expect_args` only if the arguments are good, or have an escape hatch to end execution of the caller if they're not.

Reply:
Done, I also couldn't use render :view... not sure why. It gives me the error
```
Missing template oaisys/pmh/error, oaisys/application/error with {:locale=>[:en], :formats=>[:html], :variants=>[], :handlers=>[:raw, :erb, :html, :builder, :ruby]}. Searched in: * "/home/connor/oaisys/app/views"
```

Reply:
The `formats` kind of gives away the issue – lacking any other information, it assumes that  you're looking to render an appropriately named HTML file. I'd guess this was probably part of the issue with the Application layout too.

The other thing you can see from the error is that, because the engine is namespaced, the search paths for the views is beginning with `oaisys` (it looked (in views) for a file named ` oaisys/pmh/error` with a matching format first, then fell back to `oaisys/application/error` and then errored out when it couldn't find either . I _think_ moving the file into `oaisys/pmh/` instead of `pmh` and then calling render with the right format should do it, eg) 

`render template: 'error', formats: [:xml], locals: { verb: verb, error_code: error_code, error_message: error_message }`, or probably more idiomatically:

```ruby
   respond_to do |format|
      format.xml { render template: 'error', locals: { verb: verb, error_code: error_code, error_message: error_message  } # I don't believe you need to explicitly spell out the format in this case
    end
```

which will nicely give the caller an error if they send the wrong Accepts header (https://apidock.com/rails/ActionController/MimeResponds/InstanceMethods/respond_to). This should fix the issue you ran into with the layout as well, I think?

Reply:
That resolved both of my issues! Thank you!